master/buildbot/util/git.py

```diff
@@ -270,53 +348,356 @@ def checkFeatureSupport(self):
+        yield self.step.downloadFileContentToWorker(
+            path, content, mode=mode, workdir=workdir,
+        )
+
+    @defer.inlineCallbacks
+    def download_ssh_private_key_if_needed(
+        self,
+        workdir: str,
+        download_wrapper_script: bool = False,
+    ) -> Generator[defer.Deferred[Any], Any, int]:
+        if self.ssh_private_key is None:
             return RC_SUCCESS
 
-        yield self.runRmdir(self._getSshDataPath())
+        assert (
+            isinstance(self.step, CompositeStepMixin) and
+            isinstance(self.step, GitMixin)
+        )
+
+        workdir = self._get_ssh_data_path(workdir)
+        yield self.step.runMkdir(workdir)
+
+        return_code = yield super().download_ssh_private_key_if_needed(
+            workdir=workdir,
+            download_wrapper_script=(
+                download_wrapper_script or
+                not self.step.supportsSshPrivateKeyAsEnvOption
+            ),
+        )
+        return return_code
+
+    @defer.inlineCallbacks
+    def remove_ssh_private_key_if_needed(
+        self, workdir: str,
+    ) -> Generator[defer.Deferred[Any], Any, int]:
+        if not self.did_download_ssh_private_key:
+            return RC_SUCCESS
+
+        assert isinstance(self.step, CompositeStepMixin)
+        yield self.step.runRmdir(self._get_ssh_data_path(workdir))
+        return RC_SUCCESS
+
+
+class GitServiceSSHAuth(AbstractGitSSHAuth):
+
+    def __init__(
+        self,
+        # step must implement all these types
+        service: BuildbotService,
+        ssh_private_key: IRenderable | None,
+        ssh_host_key: IRenderable | None,
+        ssh_known_hosts: IRenderable | None,
+    ) -> None:
+        self._service = service
+
+        super().__init__(ssh_private_key, ssh_host_key, ssh_known_hosts)
+
+    def adjust_command_params_for_ssh_private_key(
+        self,
+        full_command: list[str],
+        full_env: dict[str, str],
+        workdir: str,
+        git_mixin: GitMixin,
+    ) -> None:
+        self._adjust_command_params_for_ssh_private_key_impl(
+            full_command, full_env,
+            workdir=workdir, git_mixin=git_mixin,
+        )
+
+    @property
+    def _path_module(self):
+        return os.path
+
+    @property
+    def _master(self):
+        assert self._service.master is not None
+        return self._service.master
+
+    @defer.inlineCallbacks
+    def _download_file(
+        self, path: str, content: str, mode: int, workdir: str | None = None,
+    ) -> Generator[defer.Deferred[Any], Any, None]:
+        writeLocalFile(path, content, mode=mode)
+        yield defer.succeed(None)
+
+    @defer.inlineCallbacks
+    def remove_ssh_private_key_if_needed(
+        self, workdir: str,
+    ) -> Generator[defer.Deferred[Any], Any, int]:
+        if not self.did_download_ssh_private_key:
+            return RC_SUCCESS
+
+        Path(
+            self._get_ssh_private_key_path(workdir)
+        ).unlink(missing_ok=True)
```

Comment:
Mostly used pathlib here for this `missing_ok`, can replace with the `os.unlink` in a try/except if preferred.